### ISSUE TITLE
hi，兄弟们，请问微调Qwen3后输出回答老是重复是什么情况？

### ISSUE THREAD

**Issue by HongyuanLI17** [KMnO4-zx/huanhuan-chat#34]
各位大佬们，刚接触LLM这一块，我用项目中嬛嬛的数据微调了Qwen3-8B的模型，使用的超参数均是默认，也就是
`config = LoraConfig(
        task_type=TaskType.CAUSAL_LM, 
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
        inference_mode=False, # 训练模式
        r=8, # Lora 秩
        lora_alpha=32, # Lora alaph，具体作用参见 Lora 原理
        lora_dropout=0.1# Dropout 比例
    )
    model = get_peft_model(model, config)
    model.print_trainable_parameters() # 打印总训练参数

    args = TrainingArguments(
        output_dir="./output/Qwen3-8B",
        per_device_train_batch_size=4,
        gradient_accumulation_steps=4,
        logging_steps=10,
        num_train_epochs=3,
        save_steps=100, 
        learning_rate=1e-4,
        save_on_each_node=True,
        gradient_checkpointing=True
    )`
但是输出往往会出现如下情况：

> 皇上：今天向皇后请安了吗
嬛嬛：臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请安，因为臣妾从未见过皇后娘娘。臣妾今日并未向皇后娘娘请

如果你有任何建议，请务必留言告诉我，谢谢大家！

**Comment by HongyuanLI17** [comment]
我已经解决这一问题，Qwen3和Llama的分词转换形式并不一样，需要对 process_func函数进行修改，如：
`def process_func(example):
    MAX_LENGTH = 384  # Qwen对中文分词更高效，但仍需保留截断逻辑保证完整性
    input_ids, attention_mask, labels = [], [], []
    
    # Qwen3的对话格式：用<|im_start|>角色名和<|im_end|>包裹各部分内容
    # 1. 构造系统提示+用户输入部分（不计算损失）
    system_prompt = (
        f"<|im_start|>system\n"
        f"Cutting Knowledge Date: December 2023\n"
        f"Today Date: 26 Jul 2024\n\n"
        f"现在你要扮演皇帝身边的女人--甄嬛<|im_end|>\n"
    )
    user_input = f"<|im_start|>user\n{example['instruction'] + example['input']}<|im_end|>\n"
    instruction = tokenizer(system_prompt + user_input, add_special_tokens=False)
    
    # 2. 构造助手响应部分（计算损失）
    response = tokenizer(f"<|im_start|>assistant\n{example['output']}<|im_end|>", add_special_tokens=False)
    
    # 拼接输入序列：指令部分 + 响应部分（Qwen3无需额外添加pad token在末尾，截断时自然处理）
    input_ids = instruction["input_ids"] + response["input_ids"]
    attention_mask = instruction["attention_mask"] + response["attention_mask"]
    # 标签：指令部分标为-100（不参与损失计算），响应部分保留原id（参与损失计算）
    labels = [-100] * len(instruction["input_ids"]) + response["input_ids"]
    
    # 截断过长序列
    if len(input_ids) > MAX_LENGTH:
        input_ids = input_ids[:MAX_LENGTH]
        attention_mask = attention_mask[:MAX_LENGTH]
        labels = labels[:MAX_LENGTH]
    
    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }`
